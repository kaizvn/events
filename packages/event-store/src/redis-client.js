import redis from 'redis';

export const createClient = (config, { debug } = { debug: false }) => {
  const client = redis.createClient({
    ...config,
    retry_strategy: ({ attempt, total_retry_time, error, time_connected }) => {
      if (attempt > 20) {
        return
      }

      return Math.min(attempt * 100, 3000);
    }
  });

  client.on('connect', () => {
    debug && console.log('client connected');
  })

  client.on('reconnecting', ({ delay, attempt }) => {
    debug && console.log('attempting (%d) to reconnect in %dms', attempt, delay);
  })

  process.on('exit', () => {
    debug && console.log('client is quiting forcefully')
    client.end(true);
  })

  return client;
};

const defaultClient = createClient({
  url: process.env.REDIS_URL
});

export default defaultClient;

export type RedisClientType = typeof defaultClient;
