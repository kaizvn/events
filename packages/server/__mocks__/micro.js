const micro = jest.genMockFromModule('micro');

module.exports = micro;
module.exports.json = async (req) => {
  if (typeof req.body === 'object') {
    return req.body
  }

  throw new Error('Invalid JSON');
};
