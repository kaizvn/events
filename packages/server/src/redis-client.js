import redis from 'redis';
import clui from 'clui';

let countdown = new clui.Spinner();
let started;

export const createClient = (config, { debug } = { debug: false }) => {
  const client = redis.createClient({
    ...config,
    retry_strategy: ({ attempt, total_retry_time, error, time_connected }) => {
      if (attempt > 20) {
        return;
      }

      return Math.min(attempt * 100, 3000);
    },
  });

  client.on('connect', () => {
    if (debug) {
      started = false;
      countdown.stop();
    }
  });

  client.on('reconnecting', ({ delay, attempt }) => {
    if (debug) {
      countdown.message(
        `attempting (${attempt}) to reconnect in ${delay}ms...`
      );
      if (!started) {
        started = true;
        countdown.start();
      }
    }
  });

  process.on('exit', () => {
    debug && console.log('client is quiting forcefully');
    client.end(true);
  });

  return client;
};

const defaultClient = createClient({
  url: process.env.REDIS_URL,
});

export default defaultClient;

export type RedisClientType = typeof defaultClient;
