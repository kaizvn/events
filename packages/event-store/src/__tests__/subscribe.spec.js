/* @flow */
import micro from 'micro';

import { delay } from 'awaiting';
import listen from 'test-listen';

import { createClient } from '../redis-client';
import fromURL from './fromURL';
import makeSubscribe from '../subscribe';

jest.unmock('micro');

const takeEvents = (count, stream) =>
  stream.take(count).scan((prev, next) => prev.concat(next), []).toPromise();

describe('subscribe endpoint', () => {
  it('should subscribe', async () => {
    const pubClient = createClient();
    const { service, unsubscribe } = makeSubscribe({
      namespc: 'test-sub',
      history: { size: 10 },
      burst: { time: 10, count: 10 },
    });
    const server = micro(service);
    const url = await listen(server);
    const { data$, abort } = fromURL(url);

    // 1 -> :ok
    // 2 -> first event
    // 3 -> second event
    // 4 -> abort
    const promise = takeEvents(4, data$);
    await delay(100);

    pubClient.publish('test-sub::events', '1:{"type":"first"}');
    pubClient.publish('test-sub::events', `2:{"type":"second","payload":123}`);
    await delay(100);

    abort();
    server.close();
    unsubscribe();
    const val = await promise;
    pubClient.quit();
    expect(val).toMatchSnapshot();
  });

  it('should work with 2 clients', async () => {
    const pubClient = createClient();
    const {service, unsubscribe} = makeSubscribe({
      namespc: 'test-sub',
      redis: {},
      history: { size: 10 },
      burst: { time: 10, count: 1 },
    });
    const server = micro(service);
    const url = await listen(server);

    const http1 = fromURL(url + '?client=1');
    const http2 = fromURL(url + '?client=2');
    await delay(100);

    const promise1 = takeEvents(4, http1.data$);
    const promise2 = takeEvents(3, http2.data$);

    // publish 2 events
    // client 1 will wait until the end
    // client 2 will abort early after the first event
    pubClient.publish('test-sub::events', '1:{"type":"first"}');

    await delay(10);
    http2.abort();

    await delay(10);
    pubClient.publish('test-sub::events', `2:{"type":"second"}`);

    await delay(10);
    http1.abort();

    unsubscribe();

    const val1 = await promise1;
    const val2 = await promise2;

    expect(val1.length).toBe(2);
    expect(val2.length).toBe(1);
  });
});
