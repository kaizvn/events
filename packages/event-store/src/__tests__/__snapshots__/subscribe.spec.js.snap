// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribe endpoint should subscribe 1`] = `
Array [
  ":ok

",
  "id: 1
event: INCMSG
data: {\\"id\\":1,\\"type\\":\\"first\\"}

",
  "id: 2
event: INCMSG
data: {\\"id\\":2,\\"type\\":\\"second\\",\\"payload\\":123}

",
]
`;
